README.rst
Overview
========

This is the recipe that will build the CellML API Python bindings with
all options enabled by default.  Currently, there are some limitations,
such as all dependencies required to build the CellML API must be
installed before this recipe can be used, and I don't think this will
work under Windows at the moment.


Supported options
=================

The recipe supports the following options:

api-version
    CellML API version to build.  Valid versions any versions that build
    via CMake and has Python bindings (>1.10), and must be present in
    the list of valid versions.

cmake-generator
    The generator to use.  Only the default option ``Unix Makefiles`` is
    supported, as this recipe is built on top of ``zc.recipe.cmmi`` 
    which will make use of ``make`` and ``make install``.

check-build
    Whether to check build time dependencies.  Default is off because it
    didn't detect GSL libraries even though it was installed for me.
    Same as passing ``-DCHECK_BUILD:BOOL=OFF`` to ``cmake``.

Other supported options:

    - enable-examples
    - enable-annotools
    - enable-ccgs
    - enable-celeds
    - enable-celeds-exporter
    - enable-cevas
    - enable-cis
    - enable-cuses
    - enable-gsl-integrators
    - enable-malaes
    - enable-python
    - enable-rdf
    - enable-spros
    - enable-srus
    - enable-telicems
    - enable-vacss

Please refer to the `CellML API Documentations`_ for what these options
do.

.. _CellML API Documentations: http://cellml-api.sourceforge.net/


Copyright/License information
=============================

This software is released under the MPL/GPL/LGPL licenses.

Please refer to the file ``COPYING.txt`` for detailed copyright
information, and ``docs`` directory for specific licenses that this
software is released under.


CHANGES.rst
0.8 (2022-04-27)
================

- Force a known supported CXX standard because newest gcc defaults to
  the incompatible c++17 standard.

0.7 (2019-01-23)
================

- Update distribution location.

0.6 (2018-07-31)
================

- Python 3 compatibility.

0.5 (2018-01-23)
================

- Added latest version of CellML API on github.
- Added RUNPATH into place.

0.4 (2012-11-05)
================

- Added CellML API 1.12.

0.3 (2012-10-03)
================

- Corrected typo for the TeLICeM flag.
- Added CellML API 1.11.

0.2 (2011-09-21)
================

- Fixed the issue where all the expected keys are assigned with an
  empty string even if unassigned; this caused various unwanted side
  effects.
- Fixed the issue where the location of this buildout recipe is 
  undefined.


0.1 (2011-09-21)
================

- Initial release of the CellML API Python bindings buildout.




cellml/recipe/api/tests/test_docs.py
# -*- coding: utf-8 -*-
"""
Doctest runner for 'cellml.recipe.api'.
"""
__docformat__ = 'restructuredtext'

import unittest
import zc.buildout.tests
import zc.buildout.testing

from zope.testing import doctest, renormalizing

optionflags =  (doctest.ELLIPSIS |
                doctest.NORMALIZE_WHITESPACE |
                doctest.REPORT_ONLY_FIRST_FAILURE)

def setUp(test):
    import zc.recipe.cmmi
    zc.buildout.testing.buildoutSetUp(test)

    # Install the recipe in develop mode
    zc.buildout.testing.install_develop('cellml.recipe.api', test)

    # Install any other recipes that should be available in the tests
    zc.buildout.testing.install_develop('zc.recipe.cmmi', test)

def test_suite():
    suite = unittest.TestSuite((
            doctest.DocFileSuite(
                '../README.rst',
                setUp=setUp,
                tearDown=zc.buildout.testing.buildoutTearDown,
                optionflags=optionflags,
                checker=renormalizing.RENormalizing([
                        # If want to clean up the doctest output you
                        # can register additional regexp normalizers
                        # here. The format is a two-tuple with the RE
                        # as the first item and the replacement as the
                        # second item, e.g.
                        # (re.compile('my-[rR]eg[eE]ps'), 'my-regexps')
                        zc.buildout.testing.normalize_path,
                        ]),
                ),
            ))
    return suite

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


cellml/recipe/api/__init__.py
# -*- coding: utf-8 -*-
"""Recipe api"""

import zc.recipe.cmmi

_api_info = {
    '1.10': (
        'http://sourceforge.net/projects/cellml-api/files/CellML-API-Nightly/1.10/20110913/src/cellml-api-1.10.tar.bz2/download',
        '37a2cf957e9db43e21c9e43f6ec3b17f',
    ),
    '1.11': (
        'http://sourceforge.net/projects/cellml-api/files/CellML-API-Nightly/1.11/20120418/src/cellml-api-1.11.tar.bz2/download',
        '64d608890cdf1d421dc66b933729556d',
    ),
    '1.12': (
        'http://sourceforge.net/projects/cellml-api/files/CellML-API-Nightly/1.12/20121031/src/cellml-api-1.12.tar.bz2/download',
        'fdd111622f53031f9efc8e03ed13d5a6',
    ),
    'git-20171213': (
        'https://dist.staging.physiomeproject.org/cellml-api-d7baf903.zip',
        'b6b3bdee5e7d6f9b55eab9bfcee18924',
    ),
}
latest = 'git-20171213'

def get_api_info(version=latest):
    info = _api_info.get(version, None)
    if info is None:
        raise ValueError(
            'api-version `%s` is not a supported version of CellML API.' 
            % version
        )
    return info


class Recipe(zc.recipe.cmmi.Recipe):
    """\
    CellML API recipe.

    Largely built on top of `zc.recipe.cmmi`.
    """

    option_passthru_keys = [
        'url',  # user might have their own source tarball
        'md5sum',  # and respective md5sum
        'autogen',  # XXX to be ignored because autogen is not used.
        'patch',
        'patch_options',
        'environment',
        'configure-options',  # this gets processed first
    ]

    option_cmake_map = {
        'check-build': 'off',
        'enable-examples': 'on',
        'enable-annotools': 'on',
        'enable-ccgs': 'on',
        'enable-celeds': 'on',
        'enable-celeds-exporter': 'on',
        'enable-cevas': 'on',
        'enable-cis': 'on',
        'enable-cgrs': 'on',
        'enable-cuses': 'on',
        'enable-gsl-integrators': 'on',
        'enable-malaes': 'on',
        'enable-python': 'on',
        'enable-rdf': 'on',
        'enable-spros': 'on',
        'enable-srus': 'on',
        'enable-telicems': 'on',
        'enable-vacss': 'on',
    }

    def __init__(self, buildout, name, options):
        self.original_options = options

        # build a set of modified options customized for this.
        api_options = {}
        # populate it with values that we need to passthrough

        for k in self.option_passthru_keys:
            value = options.get(k, None)
            if value is not None:
                # let parent deal with anything missing.
                api_options[k] = value

        self.api_version = options.get('api-version', None)
        self.cmake_generator = options.get('cmake-generator', 'Unix Makefiles')

        if api_options.get('url', None) is None:
            # if user did not specified a specific url we assume to
            # check for a version.
            api_options['url'], api_options['md5sum'] = get_api_info(
                self.api_version)

        api_options['source-directory-contains'] = options.get(
            'source-directory-contains', 'CMakeLists.txt')
        api_options['configure-command'] = options.get(
            'configure-command', 'cmake')

        # construct extra options
        extra_options = self.build_options(self.cmake_generator, options)
        api_options['extra_options'] = extra_options

        # continue on with the parent, with our modified options.
        super(Recipe, self).__init__(buildout, name, api_options)

        # Since we did create and pass in a completely new dictionary
        # into super class, the location value needs to be assigned
        # back into here too.
        options['location'] = api_options['location']

    def build_options(self, cmake_generator, options):
        results = []
        results.append("-G '%s'" % cmake_generator)
        for k, v in self.option_cmake_map.iteritems():
            # convert to cmake keys.
            confkey = k.replace('-', '_').upper()
            results.append('-D%s:BOOL=%s' % (confkey, options.get(k, v)))

        # construct the `extra_options` string
        return ' '.join(results)

    def cmmi(self, dest):
        if self.configure_options is None:
            # unless it got assigned earlier
            self.configure_options = \
                '-DCMAKE_INSTALL_PREFIX:PATH=%s ' \
                '-DCMAKE_INSTALL_RPATH:PATH=%s/lib:%s/lib/cgrs_modules' % (
                    dest, dest, dest)

        return super(Recipe, self).cmmi(dest)
